docs/decisions/0022-planner-telemetry-enhancement.md

```diff
@@ -0,0 +1,225 @@
+
+### Function hooks
+
+Function hooks allow developers to inject logic to kernel that will be executed before or after a function is invoked. Example use cases include logging the function input before a function is invoked, and logging results after the function returns.
+For more information, please refer to the following ADRs:
+
+1. [Kernel Hooks Phase 1](./0005-kernel-hooks-phase1.md)
+2. [Kernel Hooks Phase 2](./0018-kernel-hooks-phase2.md)
+
+This approach will automatically inject default callbacks that sends telemetry data to Application Insights to all functions when the feature is turned on.
+
+Pros:
+
+1. Maximum exposure and flexibility to the developers. i.e. App developers can very easily what additional information is needed for individual functions by adding more callbacks.
+
+Cons:
+
+1. Does not use the existing System.Diagnostic framework that has been set up in the kernel. i.e. InstrumentedSKFunction, InstrumentedPlan, and InstrumentedPlanner.
+2. Does not provide the full trace detail.
+3. Difficult to create a telemetry service agnostic solution.
+
+### Distributed tracing
+
+Distributed tracing is a diagnostic technique that can localize failures and performance bottlenecks within distributed applications. .Net has native support to add distributed tracing in your libraries and .Net libraries are also instrumented to produce distributed tracing information automatically.
+
+For more information, please refer to this document: [.Net distributed tracing](https://learn.microsoft.com/en-us/dotnet/core/diagnostics/)
+
+Overall pros:
+
+1. Native .Net support and the kernel has set this up.
+
+Overall cons:
+
+1. Less flexibility for app developers to add custom traces and metrics.
+
+#### Metrics
+
+Pros:
+
+1. Easy to use for continuous monitoring.
+2. Has been set up in SK.
+
+Cons: No additional cons identified given the scope of this document.
+
+#### Activity tags & baggage
+
+```csharp
+// In the instrumented planner
+activity?.AddBaggage("Operation", $"{PlannerType}.CreatePlan");
+
+try
+{
+  ...
+  // If planner runs successfully
+  activity?.AddTag("Status", "Success");
+
+  // If it's in debugging mode
+  // Due to privacy concerns, we need to make certain telemetry items optional.
+  activity?.AddTag("Goal", goal);
+  activity?.AddTag("Plan", plan.ToSafePlanString());
+  activity?.AddTag("ExecutionTime", duration);
+}
+catch(...)
+{
+  // If planner fails
+  activity?.AddTag("Status", "Failed");
+}
+```
+
+```csharp
+// In Connectors.AI.OpenAI ClientBase.cs
+private static readonly ActivitySource s_activitySource = new(typeof(ClientBase).FullName);
+
+// In internal GetTextResults and GetChatResults calls
+using var activity = s_activitySource.StartActivity($"{this.GetType().Name}.GetTextResults");
+if (activity?.GetBaggageItem("Operation") is not null)
+{
+    activity?.AddTag("Operation", activity.GetBaggageItem("Operation"));
+}
+```
+
+In an application that uses SK and Application Insights:
```

Comment:
I mentioned before would be good if can sync up with the observability team. I am wondering here is it necessary the app need to manually to listen to the activity events? App Insights can't be enabled to automatically do this listening and collection?

Reply:
OpenTelemetry should be used to do this automatically. And yes, AppInsights supports OpenTelemetry already, so indirectly - this is supported in ApplicationInsights without these manual listeners.

Reply:
Thanks! This is taken from our current example application. We will update that application as well!